### PR TITLE
Remove `MonoItems`.

### PR DESCRIPTION
**Pull request by nnethercote**
PR #97168 was created to address this comment in `record_accesses`:
```
// We collect this into a `SmallVec` to avoid calling `is_inlining_candidate` in the lock.
// FIXME: Call `is_inlining_candidate` when pushing to `neighbors` in `collect_items_rec`
// instead to avoid creating this `SmallVec`.
```
It removed the need for the `SmallVec` (a good thing) by introducing `MonoItems`, which introduced some additional complexity (a bad thing).

However, the reason the `SmallVec` was being used in the first place -- "to avoid calling `is_inlining_candidate` in the lock" -- isn't necessary, as far as I can tell. Perhaps it was at one point.

So this commit just moves what is left of the call to `is_inlining_candidate` -- a call the innocuous `instantiation_mode` -- inside the lock. This avoids the complexity of `MonoItems` sometimes computing inlining data and sometimes not.

r? @wesleywiser

### CODE REVIEW

**Comment by nnethercote** [comment]
cc @SparrowLii, @cjgillot

File: compiler/rustc_monomorphize/src/collector.rs [hunk 1]
```diff
@@ -287,9 +252,10 @@ impl<'tcx> InliningMap<'tcx> {
         self.targets.reserve(new_items_count);
         self.inlines.ensure(new_items_count_total);
 
-        for (i, (Spanned { node: mono_item, .. }, inlined)) in new_targets.into_iter().enumerate() {
+        for (i, Spanned { node: mono_item, .. }) in new_targets.into_iter().enumerate() {
             self.targets.push(*mono_item);
-            if *inlined {
+            let is_inlined = mono_item.instantiation_mode(tcx) == InstantiationMode::LocalCopy;
```
Comment (SparrowLii) [inline compiler/rustc_monomorphize/src/collector.rs:257]: The key lies in the cost of calling `instantiation_mode`. Calling it within the lock will increase the locking time. It's not clear to me that this has a noticeable effect on the efficiency of monomorphization. cc @Zoxc

Comment (nnethercote) [reply compiler/rustc_monomorphize/src/collector.rs:257]: The comment was added in #67756. Reviewers asked for more explanation in comments, but the PR author refused. It seems like it was for a deadlock. I ran the `ui` tests with the parallel front-end enabled without problem.

Comment (nnethercote) [reply compiler/rustc_monomorphize/src/collector.rs:257]: #112128 is an alternative approach that avoids the whole locking issue.

**Pull request closed without merging**